### PR TITLE
Feature/dip 301 - introduce AWS cloudwatch support

### PR DESCRIPTION
**Pull request by Tanemahuta**

### CODE REVIEW

**Comment by gaffa** [comment]
@ulich : can you review/merge and utilize in our dash?

File: dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import de.axelspringer.ideas.tools.dash.business.check.checkresult.CheckResult;
+import de.axelspringer.ideas.tools.dash.presentation.State;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * {@link CheckExecutor} for {@link CloudWatchCheck}.
+ * @author Christian Heike
```
Comment (CanardSauvage) [inline dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java]: We have the git history vor the author ;)

File: dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+@Service
+public class CloudWatchCheckExecutor implements CheckExecutor<CloudWatchCheck> {
+
+    @Override
+    public List<CheckResult> executeCheck(final CloudWatchCheck check) {
+        final DescribeAlarmsResult alarms = check.getAmazonCloudwatchClient().describeAlarms(check.getAlarmDescription());
```
Comment (CanardSauvage) [inline dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java]: Wouldn't it be better to build the client here? The check could have a config for the client.

File: dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutorTest.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import static org.junit.Assert.*;
+
+/**
+ * Created by Tanemahuta on 27.03.17.
+ */
+public class CloudWatchCheckExecutorTest {
```
Comment (CanardSauvage) [inline dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutorTest.java]: Remove the class if it is empty or write some tests.

File: dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    public CloudWatchCheckTest(final String sourceValue, final State expectedState) {
+        this.sourceValue = sourceValue;
+        this.expectedState = expectedState;
+    }
+
+    @Test
+    public void mapState() throws Exception {
+        // expect:
+        assertThat(CloudWatchCheck.mapState(sourceValue), equalTo(expectedState));
```
Comment (CanardSauvage) [inline dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckTest.java]: I think these Parameterized kind of tests make the test less readable. An explicit test with the values would be far easier to read and understand.

File: dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        return alarms.getMetricAlarms().parallelStream().map(v -> factorCheckResult(v, check)).collect(Collectors.toList());
+    }
+
+    private CheckResult factorCheckResult(final MetricAlarm metricAlarm, final CloudWatchCheck check) {
```
Comment (CanardSauvage) [inline dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java]: "factor" is an uncommon word for this kind of method and I think it is wrong - "to factor in" means something different. I read this method name as "take check result into consideration".

File: dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java (hunk 1 above)
Comment (gaffa) [inline dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java]: But he has a very nice name! (Im removing it)

File: dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java (hunk 2 above)
Comment (gaffa) [inline dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java]: Ill look into it after merge.

File: dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutorTest.java (hunk 3 above)
Comment (gaffa) [inline dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutorTest.java]: Ill look into it after merge.

File: dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckTest.java (hunk 4 above)
Comment (gaffa) [inline dash-core/src/test/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckTest.java]: Ill look into it after merge.

File: dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java (hunk 5 above)
Comment (gaffa) [inline dash-core/src/main/java/de/axelspringer/ideas/tools/dash/business/cloudwatch/CloudWatchCheckExecutor.java]: Ill look into it after merge.

**Comment by CanardSauvage** [comment]
btw aws-sdk increases the jar from 20 MB to 70 MB.

**Comment by gaffa** [comment]
Damn. Thats the reason why we didnt use sdks for the other integrations. What would you do? Try to work around the sdk/integrate directly with the api, create a separate module/dependency for cloudwatch or simply live with it?

**Comment by CanardSauvage** [comment]
For now: Live with it or <dependency><optional>true</optional><dependency> and add some infos in the README how to use.

As we run the dash on VMs with enough RAM this is not that of a problem.

But for the longterm: ideas-dash-core and submodules like ideas-dash-aws, which you can add as additional plugins.

**Pull request merged**